File: be/src/http/action/debug_point_action.cpp
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

#include "http/action/debug_point_action.h"

#include "common/config.h"
#include "http/http_channel.h"
#include "http/http_status.h"
#include "util/debug_points.h"
#include "util/time.h"

namespace doris {

void BaseDebugPointAction::handle(HttpRequest* req) {
    LOG(INFO) << "accept one request " << req->debug_string();
    Status status;
    if (config::enable_debug_points) {
        status = _handle(req);
    } else {
        status = Status::InternalError(
                "Disable debug points. please check config::enable_debug_points");
    }
    std::string result = status.to_json();
    LOG(INFO) << "handle request result:" << result;
    if (status.ok()) {
        HttpChannel::send_reply(req, HttpStatus::OK, result);
    } else {
        HttpChannel::send_reply(req, HttpStatus::INTERNAL_SERVER_ERROR, result);
    }
}

Status AddDebugPointAction::_handle(HttpRequest* req) {
    std::string name = req->param("debug_point");
    std::string execute = req->param("execute");
    std::string timeout = req->param("timeout");
    if (name.empty()) {
        return Status::InternalError("Empty debug point name");
    }
    auto debug_point = std::make_shared<DebugPoint>();
    try {
        if (!execute.empty()) {
            debug_point->execute_limit = std::stol(execute);
        }
    } catch (const std::exception& e) {
        return Status::InternalError("Invalid execute limit format, execute {}, err {}", execute,
                                     e.what());
    }
    try {
        if (!timeout.empty()) {
            int64_t timeout_second = std::stol(timeout);
            if (timeout_second > 0) {
                debug_point->expire_ms = MonotonicMillis() + timeout_second * MILLIS_PER_SEC;
            }
        }
    } catch (const std::exception& e) {
        return Status::InternalError("Invalid timeout format, timeout {}, err {}", timeout,
                                     e.what());
    }

    debug_point->params = *(req->params());

    DebugPoints::instance()->add(name, debug_point);

    return Status::OK();
}

Status RemoveDebugPointAction::_handle(HttpRequest* req) {
    std::string debug_point = req->param("debug_point");
    if (debug_point.empty()) {
        return Status::InternalError("Empty debug point name");
    }

    DebugPoints::instance()->remove(debug_point);

    return Status::OK();
}

Status ClearDebugPointsAction::_handle(HttpRequest* req) {
    DebugPoints::instance()->clear();

    return Status::OK();
}

} // namespace doris

File: fe/fe-core/src/main/java/org/apache/doris/httpv2/rest/DebugPointAction.java
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

package org.apache.doris.httpv2.rest;

import org.apache.doris.common.Config;
import org.apache.doris.common.util.DebugPointUtil;
import org.apache.doris.httpv2.entity.ResponseEntityBuilder;
import org.apache.doris.mysql.privilege.PrivPredicate;
import org.apache.doris.qe.ConnectContext;

import com.google.common.base.Strings;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestMethod;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;

/**
 * This class is used to add, remove or clear debug points.
 */
@RestController
public class DebugPointAction extends RestBaseController {

    @RequestMapping(path = "/api/debug_point/add/{debugPoint}", method = RequestMethod.POST)
    protected Object addDebugPoint(@PathVariable("debugPoint") String debugPoint,
            @RequestParam(name = "execute", required = false, defaultValue = "") String execute,
            @RequestParam(name = "timeout", required = false, defaultValue = "") String timeout,
            HttpServletRequest request, HttpServletResponse response) {
        if (!Config.enable_debug_points) {
            return ResponseEntityBuilder.internalError(
                    "Disable debug points. please check Config.enable_debug_points");
        }
        executeCheckPassword(request, response);
        checkGlobalAuth(ConnectContext.get().getCurrentUserIdentity(), PrivPredicate.ADMIN);
        if (Strings.isNullOrEmpty(debugPoint)) {
            return ResponseEntityBuilder.badRequest("Empty debug point name.");
        }
        int executeLimit = -1;
        if (!Strings.isNullOrEmpty(execute)) {
            try {
                executeLimit = Integer.valueOf(execute);
            } catch (Exception e) {
                return ResponseEntityBuilder.badRequest(
                        "Invalid execute format: " + execute + ", err " + e.getMessage());
            }
        }
        long timeoutSeconds = -1;
        if (!Strings.isNullOrEmpty(timeout)) {
            try {
                timeoutSeconds = Long.valueOf(timeout);
            } catch (Exception e) {
                return ResponseEntityBuilder.badRequest(
                        "Invalid timeout format: " + timeout + ", err " + e.getMessage());
            }
        }
        DebugPointUtil.addDebugPoint(debugPoint, executeLimit, timeoutSeconds);
        return ResponseEntityBuilder.ok();
    }

    @RequestMapping(path = "/api/debug_point/remove/{debugPoint}", method = RequestMethod.POST)
    protected Object removeDebugPoint(@PathVariable("debugPoint") String debugPoint,
            HttpServletRequest request, HttpServletResponse response) {
        if (!Config.enable_debug_points) {
            return ResponseEntityBuilder.internalError(
                    "Disable debug points. please check Config.enable_debug_points");
        }
        executeCheckPassword(request, response);
        checkGlobalAuth(ConnectContext.get().getCurrentUserIdentity(), PrivPredicate.ADMIN);
        if (Strings.isNullOrEmpty(debugPoint)) {
            return ResponseEntityBuilder.badRequest("Empty debug point name.");
        }
        DebugPointUtil.removeDebugPoint(debugPoint);
        return ResponseEntityBuilder.ok();
    }

    @RequestMapping(path = "/api/debug_point/clear", method = RequestMethod.POST)
    protected Object clearDebugPoints(HttpServletRequest request, HttpServletResponse response) {
        if (!Config.enable_debug_points) {
            return ResponseEntityBuilder.internalError(
                    "Disable debug points. please check Config.enable_debug_points");
        }
        executeCheckPassword(request, response);
        checkGlobalAuth(ConnectContext.get().getCurrentUserIdentity(), PrivPredicate.ADMIN);
        DebugPointUtil.clearDebugPoints();
        return ResponseEntityBuilder.ok();
    }
}
